Fix `importers[path] is not a function` in yarn workspaces

kazuma1989 opened the pull request:
Hello, maintainers! This is a wonderful package. 😄 

I have recently implemented it in my project.
It was working well, but when I enabled `features.storyStoreV7`, I could not see the preview.
I get an error in the iframe saying `importers[path] is not a function`.

<img width="952" alt="image" src="https://user-images.githubusercontent.com/15844862/144958627-6f1e18b6-5ef9-41d6-aa55-28e9c82a832c.png">

The reason is that my project is a monorepo using yarn workspaces.

I've added `example-workspaces` to reproduce the issue.
You can see it by checking out the commit `5d3bcc81bc6c12132ad3c6a16be898b9caa95bb5` and running the following command.

```
cd packages/example-workspaces
yarn run storybook
```

Modifying `codegen-importfn-script.js` has resolved the issue.
The `example-react` which is not a yarn workspaces is also working fine.

Thanks.

joshwooding commented on packages/storybook-builder-vite/codegen-importfn-script.js:
```diff
@@ -16,7 +16,15 @@ const { normalizePath } = require('vite');
 async function toImportFn(stories) {
     const objectEntries = stories
-        .map(file => `  './${normalizePath(path.relative(process.cwd(),file))}': async () => import('/@fs/${file}')`);
+        .map(file => {
```
Wondering if we should do something like https://github.com/storybookjs/storybook/blob/d42f89a9d43cb7753e5484c9dd2bcf8677053239/lib/core-server/src/utils/watch-story-specifiers.ts#L21-L23

Having identical logic to Storybook is more favourable in my opinion. Something like:

```js
// Paths get passed either with no leading './' - e.g. `src/Foo.stories.js`,
// or with a leading `../` (etc), e.g. `../src/Foo.stories.js`.
// We want to deal in importPaths relative to the working dir, so we normalize
// Taken from https://github.com/storybookjs/storybook/blob/d42f89a9d43cb7753e5484c9dd2bcf8677053239/lib/core-server/src/utils/watch-story-specifiers.ts#L21-L23
function toImportPath(relativePath) {
    return relativePath.startsWith('.') ? relativePath : `./${relativePath}`;
}

...

async function toImportFn(stories) {
    const objectEntries = stories
        .map(file => `  './${toImportPath(normalizePath(path.relative(process.cwd(),file)))}': async () => import('/@fs/${file}')`);


```

joshwooding commented:
A minor thing but I don't think we need app1 and app2? I think app1 would suffice?

IanVS commented on packages/storybook-builder-vite/codegen-importfn-script.js:
```diff
@@ -16,7 +16,15 @@ const { normalizePath } = require('vite');
 async function toImportFn(stories) {
     const objectEntries = stories
-        .map(file => `  './${normalizePath(path.relative(process.cwd(),file))}': async () => import('/@fs/${file}')`);
+        .map(file => {
```
I'm not sure the naive approach of `toImportPath` works here, since I think it fails for files and folders starting with a `.` (such as hidden files/folders like `.storybook`).  

`normalizePath` from vite is a wrapper around `path.posix.normalize` which handles windows paths (https://github.com/vitejs/vite/blob/b40163c794b1c3be3f1715ed8a04d03dc72388b4/packages/vite/src/node/utils.ts#L83`).  I believe that `path.normalize` will never include `./` or `../` at the start, so I'm not sure what the change here is attempting to solve.  @kazuma1989 can you explain why you're checking for relative paths like this?

IanVS replied:
ah hah, that makes sense.  So I guess we'd just need to check for `../`, and only add `./` if it's not there, right?

joshwooding replied:
Yep :) This fix is fine. Just wanted to see if there was Storybook code we could copy but seems like there isn't :(

IanVS replied:
Technically the `!normalizedPath.startsWith('./') &&` part of this fix is not necessary, right?

joshwooding replied:
Good point. It's not needed :)

kazuma1989 replied:
Thank you so much, @joshwooding and @IanVS!

Yes, @joshwooding gave a perfect explanation.
And the point @IanVS made is also right. Checking `./` is not needed. I added that part just because I didn't know `normalizePath` returns a leading `./` or not.

Fixed in https://github.com/eirslett/storybook-builder-vite/pull/172/commits/8331490662708a9f4cba4da280968790285264b1

kazuma1989 commented:
> A minor thing but I don't think we need app1 and app2? I think app1 would suffice?

Yes. I added app2 to show a little complex example. It's not the essence for the fix. 👍 

Fixed in https://github.com/eirslett/storybook-builder-vite/pull/172/commits/029e34195e546f0b6cc7395138607213675fe1b7

joshwooding commented:
Thank you @kazuma1989 🎉
